### PR TITLE
Use cubic Bézier curve to sample NACA 4-digit airfoil

### LINKED ISSUE #320: Use Bezier curve to fit NACA 4-digit airfoil
In the current prototype of the `naca` app, the airfoil is described [using sampled edges](https://github.com/solvcon/modmesh/blob/bc9e26f07585d74b2fb4d96c194ff209e7c78996/modmesh/app/naca.py#L104).  I'd like to try to describe the airfoil using Bezier curves, so that generating meshes for the airfoil shapes ca use arbitrarily high resolution.

### Progress

- [x] PR #384 adds a prototype that samples the 4th-order polynomial profile of the airfoil using both cubic Bezier curves and line segments, but has not done fitting.  More details in the comment: https://github.com/solvcon/modmesh/issues/320#issuecomment-2281235786
- [x] PR #406 refactors the code added in PR #384.  More details in the comment: https://github.com/solvcon/modmesh/issues/320#issuecomment-2282744365
- [ ] Use cubic Bezier curves and line segments to fit the airfoil profile.

### Open for discussions

Add GUI control for changing the airfoil drawing and fitting parameters.

### PR DESCRIPTION
**Pull request by YenPeiChen07**
@yungyuc, 

This PR is for using cubic Bézier curve to sample NACA 4-digit airfoil.

The control points cubic Bézier curve are determined by two consecutive data points and two computed points derived from these two consecutive points.

Currently, the result looks fine but I am not sure if it meets our desired accuracy. I think I need an error metric to optimize the control points for better results or to provide a quantitive analysis.

Result:
image

### CODE REVIEW

File: modmesh/gui/naca_bezier.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
+# Copyright (c) 2021, Yung-Yu Chen <yyc@solvcon.net>
```
Comment (yungyuc) [inline modmesh/gui/naca_bezier.py]: We should not use a separate file for the NACA airfoil.  Please directly add the cubic-Bezier-curve approximation in the existing `naca.py` file without destroying its original capabilities.  That is, the updated `naca.py` should approximate the airfoil using piece-wise linear segments and piece-wise cubic-Bezier curves.

**Review by yungyuc — CHANGES_REQUESTED** [review]
This is a good start.  Please fix the basic issue:

- [x] Use the existing `naca.py` to hold the new feature of cubic-Bezier-curve approximation.

---

@YenPeiChen07 I agree that we need a metric for determining the cubic-Bezier-curve approximation.  Could you please try to use the basic least-square (L2 norm)?

File: modmesh/gui/naca_bezier.py (hunk 1 above)
Comment (YenPeiChen07) [reply modmesh/gui/naca_bezier.py]: Ok, I will fit it.

**Comment by YenPeiChen07** [comment]
Sure, I can try to use L2-norm as metric and use the exiting file `naca.py` to hold the new feature of cubic-Bezier-curve approximation.

I should be able to finish it by Tuesday.

**Comment by YenPeiChen07** [comment]
Late update to my plan.

Currently, the airfoil can be drawn in both linear approximation and cubic bezier approximation in the existing file `naca.py`.

The next try is to calculate the L2-norm as metric for determining cubic Bézier curve approximation.

Details for next step:
1.  Get the x-axis value of sampled points on each cubic Bézier curve (pX)
2. Get the y-axis value (pY1) based on sampled point (pX)
3. Get the y-axis value from NACA parametric curve (pY2)
4. Calculate L2-norm between pY1 and pY2. (The square root of the sum of squares of the residual between pY1 and pY2)

File: modmesh/gui/naca.py [hunk 2]
```diff
@@ -94,16 +93,49 @@ def calc_naca4_points(number, npoint, open_trailing_edge=False,
...
+
+        b = world.add_bezier(
+            [
+                Vector(points[it, 0], points[it, 1], 0),
+                Vector(p1[0], p1[1], 0),
+                Vector(p2[0], p2[1], 0),
+                Vector(points[it + 1, 0], points[it + 1, 1], 0),
+            ]
+        )
+        b.sample(7)
```
Comment (yungyuc) [inline modmesh/gui/naca.py]: Number of sample points should be made as an argument.  The default should be `None`, means automatic determination of the number of linear sample points.  You should come up with a formula to determine a reasonable number of linear sample points.

File: modmesh/gui/naca.py [hunk 3]
```diff
@@ -94,16 +93,49 @@ def calc_naca4_points(number, npoint, open_trailing_edge=False,
     return ret
 
 
-def draw_naca4(world, number, npoint, fac, off_x, off_y, **kw):
+def linear_approximation(world, points):
```
Comment (yungyuc) [inline modmesh/gui/naca.py]: Rename `linear_approximation` to `linear_approximate`.

File: modmesh/gui/naca.py [hunk 4]
```diff
@@ -94,16 +93,49 @@ def calc_naca4_points(number, npoint, open_trailing_edge=False,
...
+    for it in range(points.shape[0] - 1):
+        e = world.add_edge
+        (points[it, 0], points[it, 1], 0, points[it + 1, 0], points[it + 1, 1], 0)
+        print(f"{it}: {e}")
+    return world.nedge
+
+
+def cubic_bezier_approximation(world, points):
```
Comment (yungyuc) [inline modmesh/gui/naca.py]: Rename `cubic_bezier_approximation` to `cubic_bezier_approximate`.

File: modmesh/gui/naca.py [hunk 5]
```diff
@@ -94,16 +93,49 @@ def calc_naca4_points(number, npoint, open_trailing_edge=False,
...
+        )
+        b.sample(7)
+
+        print("control point 1: ", p1)
+        print("control point 2: ", p2)
+        print("bezier locus: ", b.nlocus)
+    return world.nbezier
+
+
+def draw_naca4(world, number, npoint, fac, off_x, off_y, use_bezier, **kw):
```
Comment (yungyuc) [inline modmesh/gui/naca.py]: Give `use_bezier` a default value of `False`.

Comment (yungyuc) [reply modmesh/gui/naca.py]: Open discussion: does it make sense to turn this function `draw_naca4` a class?

File: modmesh/gui/naca.py [hunk 6]
```diff
@@ -35,8 +35,7 @@
 from modmesh import view
 
 
-def calc_naca4_points(number, npoint, open_trailing_edge=False,
-                      cosine_spacing=False):
+def calc_naca4_points(number, npoint, open_trailing_edge=False, cosine_spacing=False):
```
Comment (yungyuc) [inline modmesh/gui/naca.py]: Why changing the format of this line?

**Review by yungyuc — CHANGES_REQUESTED** [review]
Things need to be changed:

- [x] Please please the linter.
- [x] Rename `linear_approximation` to `linear_approximate`, `cubic_bezier_approximation` to `cubic_bezier_approximate` (use a verb in the function name).
- [x] Add an argument to `cubic_bezier_approximate` for the number of linear sample points.
- [x] Give the default value of `False` to the argument `use_bezier` to `draw_naca4`.

----

Discussions:

- [x] Does it make sense to turn the function `draw_naca4` into a class?

----

> Late update to my plan.
> 
> Currently, the airfoil can be drawn in both linear approximation and cubic bezier approximation in the existing file `naca.py`.

This is the focus of this PR.  

> The next try is to calculate the L2-norm as metric for determining cubic Bézier curve approximation.
> 
> Details for next step:
> 
> 1. Get the x-axis value of sampled points on each cubic Bézier curve (pX)
> 2. Get the y-axis value (pY1) based on sampled point (pX)
> 3. Get the y-axis value from NACA parametric curve (pY2)
> 4. Calculate L2-norm between pY1 and pY2. (The square root of the sum of squares of the residual between pY1 and pY2)

Please create an issue to track the coding work for the linear regression.  It should use a different PR.

File: modmesh/gui/naca.py (hunk 2 above)
Comment (YenPeiChen07) [reply modmesh/gui/naca.py]: Does this mean that we can call `b.sample()` to sample linear sample points and there is a formula in the [function definition of `sample()`](https://github.com/YenPeiChen07/modmesh/blob/9f18780f1843f1faf2e05922be60674c3c78a353/cpp/modmesh/universe/bezier.hpp#L285) to determine a reasonable number of linear sample points?

Besides, we still can pass an argument to `sample()` function, like we do now with `b.sample(7)`.

File: modmesh/gui/naca.py (hunk 5 above)
Comment (YenPeiChen07) [reply modmesh/gui/naca.py]: Yes, I think so.  Turning function `draw_naca4` a class would be better to manage parameters or behaviors for an airfoil. From my view, `number` and `npoint` in `draw_naca4` are static information of an airfoil and others are the factors about drawing an airfoil. They can be defined in a class as variables and functions separately.

File: modmesh/gui/naca.py (hunk 6 above)
Comment (YenPeiChen07) [reply modmesh/gui/naca.py]: Because there is an error `./modmesh/gui/naca.py:38:80: E501 line too long (86 > 79 characters)` when I run `make flake8`, I change it for fix it. But it actually did not fix and turn out the other error `./modmesh/gui/naca.py:38:64: W291 trailing whitespace`. 

I will figure out how to deal with linting and formatting Python in vscode.

Comment (yungyuc) [reply modmesh/gui/naca.py]: The code after your change uses 86 characters:

```python
def calc_naca4_points(number, npoint, open_trailing_edge=False, cosine_spacing=False):
```

It seems the code before your change is OK.

File: modmesh/gui/naca.py (hunk 2 above)
Comment (yungyuc) [reply modmesh/gui/naca.py]: I mean the number of sample points (passed into `sample()`) should be made as an argument to this function `cubic_bezier_approximate()`.  When the argument is `None`, you use an algorithm to determine what is a proper value to send to `sample()`, instead of hard-coding to 7.

File: modmesh/gui/naca.py (hunk 5 above)
Comment (yungyuc) [reply modmesh/gui/naca.py]: I think doing the function-to-class refactor is OK with this PR or another.

File: modmesh/gui/naca.py (hunk 2 above)
Comment (YenPeiChen07) [reply modmesh/gui/naca.py]: Thanks for correcting me. I found I misunderstood what you meant.

It may look like:
```
def cubic_bezier_approximate(world, points, nsample=None):
    Vector = core.Vector3dFp64

    if nsample == None:
        # find a formule to determine a reasonable number of linear sample points.

    for it in range(points.shape[0] - 1):
       ...
       b.sample(nsample)

return world.nbezier
```

Comment (yungyuc) [reply modmesh/gui/naca.py]: That is what I meant.

File: modmesh/gui/naca.py [hunk 7]
```diff
@@ -94,6 +94,19 @@ def calc_naca4_points(number, npoint, open_trailing_edge=False,
...
 
 
+def cal_sample_points(points):
+    # this is to determine the numbe of sample points for a cubic bezier curve
+    spacing = 0.01
+    samples = [-1] * (points.shape[0] - 1)
+    for it in range(points.shape[0] - 1):
+        length = np.sqrt(np.power(points[it, 0] - points[it+1, 0], 2) +
+                         np.power(points[it, 1] - points[it+1, 1], 2))
+        segments = int((length / spacing))
```
Comment (YenPeiChen07) [inline modmesh/gui/naca.py]: Determine how many segments needed for a cubic Bézier curve by dividing the length with spacing.

File: modmesh/gui/naca.py [hunk 8]
```diff
@@ -94,6 +94,19 @@ def calc_naca4_points(number, npoint, open_trailing_edge=False,
...
+        nsample = segments - 1 if segments > 2 else 2  # the min nsampe is 2
```
Comment (YenPeiChen07) [inline modmesh/gui/naca.py]: Determine number of sample points based on number of segments. The minimum sample point is 2.

**Comment by YenPeiChen07** [comment]
@yungyuc,

The new commit is to fulfill the requests you mentioned before:

> Things need to be changed:
> 
> * [x]  Please please the linter.
> * [x]  Rename `linear_approximation` to `linear_approximate`, `cubic_bezier_approximation` to `cubic_bezier_approximate` (use a verb in the function name).
> * [x]  Add an argument to `cubic_bezier_approximate` for the number of linear sample points.
> * [x]  Give the default value of `False` to the argument `use_bezier` to `draw_naca4`.


To-do: 
- [ ] (refactor) Turn function `draw_naca4` a class

The function-to-class refactor for `draw_naca4` will be done soon. I would like to finish the linear regression first.

File: modmesh/gui/naca.py [hunk 9]
```diff
@@ -94,16 +94,68 @@ def calc_naca4_points(number, npoint, open_trailing_edge=False,
...
+        b = world.add_bezier(
+            [
+                Vector(points[it, 0], points[it, 1], 0),
+                Vector(p1[0], p1[1], 0),
+                Vector(p2[0], p2[1], 0),
+                Vector(points[it + 1, 0], points[it + 1, 1], 0),
+            ]
+        )
+        b.sample(nsample[it])
+        print("point 1: ", points[it])
```
Comment (yungyuc) [inline modmesh/gui/naca.py]: Remove the debug code.

File: modmesh/gui/naca.py [hunk 10]
```diff
@@ -94,16 +94,68 @@ def calc_naca4_points(number, npoint, open_trailing_edge=False,
     return ret
 
 
-def draw_naca4(world, number, npoint, fac, off_x, off_y, **kw):
+def cal_sample_points(points):
```
Comment (yungyuc) [inline modmesh/gui/naca.py]: A convention in modmesh is to use `calc` as a shorthand for `calculate`.

File: modmesh/gui/naca.py (hunk 7 above)
Comment (yungyuc) [reply modmesh/gui/naca.py]: Use [`//` (floored quotient / integer division)](https://docs.python.org/3/library/stdtypes.html) instead.

File: modmesh/gui/naca.py [hunk 11]
```diff
@@ -94,16 +94,68 @@ def calc_naca4_points(number, npoint, open_trailing_edge=False,
...
+    # this is to determine the numbe of sample points for a cubic bezier curve
+    spacing = 0.01
+    samples = [-1] * (points.shape[0] - 1)
+    for it in range(points.shape[0] - 1):
```
Comment (yungyuc) [inline modmesh/gui/naca.py]: You can use the [batch hypotenuse function](https://numpy.org/doc/stable/reference/generated/numpy.hypot.html) to avoid the loop.

File: modmesh/gui/naca.py [hunk 12]
```diff
@@ -112,10 +164,20 @@ def runmain():
     A simple example for drawing a couple of cubic Bezier curves.
     """
     w = core.WorldFp64()
-    draw_naca4(w, number='0012', npoint=101, fac=5.0, off_x=0.0, off_y=2.0,
-               open_trailing_edge=False, cosine_spacing=True)
+    draw_naca4(
+        w,
```
Comment (yungyuc) [inline modmesh/gui/naca.py]: This one-argument-a-line style takes too many line.  Please keep the original terse style.

**Review by yungyuc — CHANGES_REQUESTED** [review]

**Comment by YenPeiChen07** [comment]
> This PR is taking a while. You may use another PR for refactoring the code into a class.
> 
> Please fix the following:
> 
> * [ ]  Use [`//` (floored quotient / integer division)](https://docs.python.org/3/library/stdtypes.html) instead of `int(...)`.
> * [ ]  Use the [batch hypotenuse function](https://numpy.org/doc/stable/reference/generated/numpy.hypot.html) to avoid the loop to calculate the number of segments.
> * [ ]  This one-argument-a-line style takes too many line.  Please keep the original terse style.

Thanks @yungyuc, I will fix them before this Thursday.

**Comment by YenPeiChen07** [comment]
@yungyuc, I have fix the problems you mentioned. Please help me take a look when you are available.

File: modmesh/gui/naca.py [hunk 13]
```diff
@@ -94,16 +94,53 @@ def calc_naca4_points(number, npoint, open_trailing_edge=False,
     return ret
 
 
-def draw_naca4(world, number, npoint, fac, off_x, off_y, **kw):
+def calc_sample_points(points):
+    # determine the number of sample points for a cubic bezier curve
+    spacing = 0.01
+    segments = np.hypot((points[:-1, 0] - points[1:, 0]),
```
Comment (yungyuc) [inline modmesh/gui/naca.py:100]: This is much cleaner.  Wonderful.

File: modmesh/gui/naca.py [hunk 14]
```diff
@@ -112,10 +149,11 @@ def runmain():
     A simple example for drawing a couple of cubic Bezier curves.
     """
     w = core.WorldFp64()
-    draw_naca4(w, number='0012', npoint=101, fac=5.0, off_x=0.0, off_y=2.0,
-               open_trailing_edge=False, cosine_spacing=True)
+    draw_naca4(w, number="0012", npoint=101, fac=5.0, off_x=0.0, off_y=2.0,
```
Comment (yungyuc) [inline modmesh/gui/naca.py]: It is not necessary to replace the single quote `'` with double quote `"`.  Please revert the unnecessary change.

File: modmesh/gui/naca.py [hunk 15]
```diff
@@ -112,10 +149,11 @@ def runmain():
...
+               open_trailing_edge=False, use_bezier=True, cosine_spacing=True)
     wid = view.mgr.add3DWidget()
     wid.updateWorld(w)
     wid.showMark()
 
+
```
Comment (yungyuc) [inline modmesh/gui/naca.py]: It is not necessary to add the additional blank line.  Please revert this.

File: modmesh/gui/naca.py [hunk 16]
```diff
@@ -94,16 +94,53 @@ def calc_naca4_points(number, npoint, open_trailing_edge=False,
...
+                        (points[:-1, 1] - points[1:, 1])) // spacing
+    nsample = np.where(segments > 2, segments - 1, 2).astype(int)
+    return nsample
+
+
+def linear_approximate(world, points):
+    for it in range(points.shape[0] - 1):
+        e = world.add_edge(points[it, 0], points[it, 1], 0,
+                           points[it + 1, 0], points[it + 1, 1], 0)
+        print(f"{it}: {e}")
```
Comment (yungyuc) [inline modmesh/gui/naca.py]: Please help remove debug prints.  The code is being improved and it's time to get rid of them.

File: modmesh/gui/naca.py [hunk 17]
```diff
@@ -94,16 +94,53 @@ def calc_naca4_points(number, npoint, open_trailing_edge=False,
...
     crds[:, 1] += off_y  # offset in y
-    for it in range(crds.shape[0] - 1):
-        e = world.add_edge(crds[it, 0], crds[it, 1], 0,
-                           crds[it + 1, 0], crds[it + 1, 1], 0)
-        print(f"{it}: {e}")
-    print("nedge:", world.nedge)
+
+    if use_bezier:
+        nbezier = cubic_bezier_approximate(world, crds)
+        print("nbezier:", nbezier)
```
Comment (yungyuc) [inline modmesh/gui/naca.py]: Please help remove debug prints.

File: modmesh/gui/naca.py [hunk 18]
```diff
@@ -94,16 +94,53 @@ def calc_naca4_points(number, npoint, open_trailing_edge=False,
...
+    else:
+        nedge = linear_approximate(world, crds)
+        print("nedge:", nedge)
```
Comment (yungyuc) [inline modmesh/gui/naca.py]: Please help remove debug prints.

**Review by yungyuc — CHANGES_REQUESTED** [review]
> @yungyuc, I have fix the problems you mentioned. Please help me take a look when you are available.

Thanks, this is much better.  Please fix the minor issues below.

- [x] It is not necessary to replace the single quote `'` with double quote `"`.  Please revert the unnecessary change.
- [x] It is not necessary to add the additional blank line. Please revert this.
- [x] This is a simple change and does not need many checkins.  Please squash the changesets into one, and provide a good summary in the commit log.
- [x] Help review and remove all debug prints.

**Comment by YenPeiChen07** [comment]
> > @yungyuc, I have fix the problems you mentioned. Please help me take a look when you are available.
> 
> Thanks, this is much better. Please fix the minor issues below.
> 
> * [ ]  It is not necessary to replace the single quote `'` with double quote `"`.  Please revert the unnecessary change.
> * [ ]  It is not necessary to add the additional blank line. Please revert this.
> * [ ]  This is a simple change and does not need many checkins.  Please squash the changesets into one, and provide a good summary in the commit log.
> * [ ]  Help review and remove all debug prints.

@yungyuc, I have removed the unnecessary changes, blank line, and debug prints. Please help review them when you have time.

**Review by yungyuc — APPROVED** [review]
> @yungyuc, I have removed the unnecessary changes, blank line, and debug prints. Please help review them when you have time.

LGTM.  Thanks.

**Comment by yungyuc** [comment]
@YenPeiChen07 Could you please attach snapshots to summarize the straight lines and cubic Bezier curves that sample the airfoil profile?
